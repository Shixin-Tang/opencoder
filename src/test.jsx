import React, { useState, useEffect } from "react"
import { render, Box, Text } from "ink2"
import Static from "./components/static.js"
import { render as renderInk } from "../test/testutil"

const Example = () => {
  const [tests, setTests] = useState([])

  useEffect(() => {
    let completedTests = 0
    let timer

    const run = () => {
      // Fake 10 completed tests
      if (completedTests++ < 10) {
        setTests((previousTests) => [
          ...previousTests,
          {
            id: previousTests.length,
            title: `Test #${previousTests.length + 1}`,
          },
        ])

console.log(renderInk(<Text color="green">✔Hello</Text>).lastFrame())

        timer = setTimeout(run, 100)
      }
    }

    run()

    return () => {
      clearTimeout(timer)
    }
  }, [])

  return (
    <Box flexDirection="column" gap={1} position="relative">
      {/* This part will be rendered once to the terminal */}
      <Static items={tests}>
        {(test) => (
          <Box key={test.id}>
            <Text color="green">✔ {test.title}</Text>
          </Box>
        )}
      </Static>

      {/* This part keeps updating as state changes */}
      <Box marginTop={1}>
        <Text dimColor>Completed tests: {tests.length}</Text>
      </Box>
    </Box>
  )
}

render(<Example />)
