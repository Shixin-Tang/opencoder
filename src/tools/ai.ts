import type { ReactNode } from "react"
import { type JSONValue, type Tool, type ToolInvocation } from "ai"

// TODO: annotate message with tool output params, eg file patch, useful for persistent chat messages.

type Streamable = ReactNode | Promise<ReactNode> | JSONValue

export type Renderer<T extends Array<any>> = (...args: T) => AsyncGenerator<Streamable, void, void>

export type CoderTool = Omit<Tool, "execute"> & {
  generate?: Renderer<any>
  renderTitle?: (part: ToolInvocation) => ReactNode
  render?: (part: ToolInvocation) => ReactNode
  execute?: Tool["execute"]
}

export function defineTool(params: CoderTool): CoderTool {
  return params
}
