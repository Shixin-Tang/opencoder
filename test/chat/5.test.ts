import { tool } from "ai"
import { convertArrayToReadableStream, MockLanguageModelV1 } from "ai/test"

import { z } from "zod"
import { waitNextRender } from "../utils/render.js"
import { setupTestEnvironment } from "./util.js"

// Define expected strings for assertions
const toolFailureErrorMessage =
  "Error executing tool failing_tool: Tool execution failed spectacularly!"
const inputPromptIndicator = ">"

test("handling tool execution error", async () => {
  // 1. Setup: Model requests a tool, tool execution fails
  const mockFailingTool = vi
    .fn()
    .mockRejectedValue(new Error("Tool execution failed spectacularly!"))

  const tools = {
    failing_tool: tool({
      execute: mockFailingTool,
      description: "A tool designed to fail",
      parameters: z.object({ param: z.string() }),
    }),
  }

  const mockModel = new MockLanguageModelV1({})
  const doStreamMock = vi.fn()
  mockModel.doStream = doStreamMock

  // Model requests the failing tool
  doStreamMock.mockImplementationOnce(async () => ({
    stream: convertArrayToReadableStream([
      {
        type: "tool-call",
        toolCallId: "tool_fail_1",
        toolName: "failing_tool",
        args: JSON.stringify({ param: "someValue" }),
      },
      {
        type: "finish",
        finishReason: "tool-calls",
        usage: { promptTokens: 5, completionTokens: 5 },
      },
    ]),
    rawCall: { rawPrompt: null, rawSettings: {} },
    type: "finish",
    finishReason: "tool-calls",
    usage: { promptTokens: 5, completionTokens: 5 },
  }))

  // Model should *not* be called again after tool failure in this flow
  // (The error message is generated client-side)

  const { instance, stdin, stdout, fiber, config } = await setupTestEnvironment({
    model: mockModel,
    customTools: tools,
  })

  // Check initial state - should only have input prompt
  const initialOutput = stdout.get()
  expect(initialOutput).includes(inputPromptIndicator)
  expect(initialOutput).not.includes(toolFailureErrorMessage)

  assert(stdin)

  // 2. Simulate user input
  stdin.emit("input", "Run the failing tool")
  await waitNextRender()
  stdin.emit("input", "\r")
  await waitNextRender()
  await waitNextRender()
  await waitNextRender()

  // Restore waitFor to allow the app time to process the tool call and handle the rejection
  await vi.waitFor(
    () => {
      const output = stdout.get()
      // Check that the specific error message IS present
      expect(output).includes(toolFailureErrorMessage)
      // Check that the input prompt reappears
      expect(output).includes(inputPromptIndicator)
    },
    { timeout: 3000 },
  )

  // 5. Verify tool was called
  expect(mockFailingTool).toHaveBeenCalledTimes(1)
  expect(mockFailingTool.mock.calls[0][0]).toEqual({ param: "someValue" })

  // 6. Verify model was NOT called a second time
  expect(doStreamMock).toHaveBeenCalledTimes(1)

  // 7. Final state check - already partially done in waitFor, but can re-verify
  const finalOutput = stdout.get()
  expect(finalOutput).includes(toolFailureErrorMessage)
  expect(finalOutput).includes(inputPromptIndicator)

  // 8. Ensure app is stable and ready for next input
  stdin.emit("input", "next message")
  await waitNextRender()
  expect(stdout.get()).toContain("next message")

  instance.unmount()
})
