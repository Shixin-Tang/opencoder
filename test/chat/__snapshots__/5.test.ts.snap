// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handling tool execution error > tool error - final state 1`] = `
"> Run the failing tool
▼  Custom tool: failing_tool
 ╭─────────────────────────────────────────────────────────────────────────────────────────────────╮
 ╰─────────────────────────────────────────────────────────────────────────────────────────────────╯

╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ An error occurred while processing your request:                                                 │
│                                                                                                  │
│ Error executing tool failing_tool: Tool execution failed spectacularly!                          │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯


╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ >                                                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
  / for commands                                                    alt + ⏎ for newline ·10 tokens"
`;

exports[`handling tool execution error > tool error - initial 1`] = `
"
╭──────────────────────────────────────────────────────────────────────────────────────────────────╮
│ >                                                                                                │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
  / for commands                                                     alt + ⏎ for newline ·0 tokens"
`;
